virtual_rainforest/soil/model.py

```diff
@@ -0,0 +1,108 @@
+
+    Describes the specific functions and attributes that the soil module should possess.
+    It's very much incomplete at the moment, and only overwrites one function in a
+    pretty basic manner. This is intended as a demonstration of how the class
+    inheritance should be handled for the model classes.
+
+    Args:
+        start_time: Point in time that the model simulation should be started.
+        end_time: Time that the model simulation should end
+        update_interval: Time to wait between updates of the model state.
+        no_layers: The number of soil layers to be modelled.
+
+    Attributes:
+        name: Names the model that is described
+    """
+
+    name = "soil"
+
+    def __init__(self, *args: Any, **kwargs: Any):
+
+        if args[1] < 1:
+            log_and_raise(
+                "There has to be at least one soil layer in the soil model!", ValueError
+            )
+        elif not isinstance(args[1], int):
+            log_and_raise("The number of soil layers must be an integer!", TypeError)
+
+        super(SoilModel, self).__init__(*args)
+        self.no_layers = args[1]
+
+    @classmethod
+    def factory(cls, config: dict[str, Any]) -> SoilModel:
+        """Factory function to initialise the soil model.
+
+        This function unpacks the relevant information from the configuration file, and
+        then uses it to initialise the model.
+
+        Args:
+            config: The complete (and validated) virtual rainforest configuration.
+
+        Raises:
+            InitialisationError: If the information required to initialise the model
+                either isn't found, or isn't of the correct type.
+        """
+        try:
+            raw_interval = config["core"]["timing"]["update_interval"]
+            # Round raw time interval to nearest minute
+            update_interval = timedelta64(int(raw_interval * 24 * 60), "m")
+            no_layers = config["soil"]["no_layers"]
+        except KeyError as e:
+            log_and_raise(
+                f"Configuration is missing information required to initialise the soil "
+                f"model. The first missing key is {str(e)}.",
+                InitialisationError,
+            )
+        except ValueError as e:
+            log_and_raise(
+                f"Configuration types appear not to have been properly validated. This "
+                f"problem prevents initialisation of the soil model. The first instance"
+                f" of this problem is as follows: {str(e)}",
+                InitialisationError,
+            )
+
+        # TODO - Add further relevant checks on input here as they become relevant
+
+        LOGGER.info(
+            "Information required to initialise the soil model successfully extracted."
+        )
+
+        return cls(update_interval, no_layers)
+
+    # THIS IS BASICALLY JUST A PLACEHOLDER TO DEMONSTRATE HOW THE FUNCTION OVERWRITING
+    # SHOULD WORK
+    def setup(self) -> None:
```

Comment:
Keeping aside this is a placeholder, what would be the purpose of the `setup` method? Why the model cannot be setup in the `factory` method or the `__init__` method? Seems redundant to me.

Reply:
I see. I think it can be split in two - or more functions - to keep the separation of concerns clear. The question is if they should be public methods defined in `BaseModel` to be overwritten by children classes or if they can be internal, defined by each class as needed.

For the use case you describe, I think `setup` does not need to be in the `BaseModel` class and can be called internally by `factory` (or `__init__` as appropriate). The only condition is that the created `grid` need to be passed as an argument to `factory`, so the pools can be setup appropriately in each cell.

A different situation is if that setup step needs information from other models. In that case, probably it is indeed necessary some sort of `setup` method used by models to talk to each other at `t=0` to define their initial state before the simulation can start.

Reply:
I think it is the latter case. I can imagine some models might not need to talk to other models (animals?) but the setup for plants and abiotic is likely to involve discussion. So, I think having the public method which is called for all models at t=0 (and might just `pass`) works here.

Reply:
In that case I would not make it an `abstractmethod` but rather a normal one that, by default, does nothing, so models that do not need any setup can skip overwriting it.